go/porcelain/deploy.go

```diff
@@ -538,10 +542,16 @@ func (n *Netlify) uploadFile(ctx context.Context, d *models.Deploy, f *FileBundl
 			context.GetLogger(ctx).WithError(operationError).Errorf("Failed to upload file %v", f.Name)
 			apiErr, ok := operationError.(apierrors.Error)
 
-			if ok && apiErr.Code() == 401 {
-				sharedErr.mutex.Lock()
-				sharedErr.err = operationError
-				sharedErr.mutex.Unlock()
+			if ok {
+				if apiErr.Code() == 401 {
+					sharedErr.mutex.Lock()
+					sharedErr.err = operationError
+					sharedErr.mutex.Unlock()
+				}
+
+				if skipRetry && (apiErr.Code() == 400 || apiErr.Code() == 422) {
+					operationError = permanentErr
```

Comment:
I might be wrong, but I think we should wrap the existing error in a `backoff.PermanentError{}` (for which `backoff` provides the `Permanent()` function:

```suggestion
					operationError = backoff.Permanent(operationError)
```

Reply:
Ah, I think that does make sense. And then when `backoff` receives the `Permanent` error, it should stop retrying, if I understand correctly.